glowfic-dl.py

```diff
@@ -130,11 +161,75 @@ def __init__(self, stamped_urls: list[StampedURL], title: str):
         self.last_update = max((stamped_url.stamp for stamped_url in stamped_urls))
 
 
+#################
+##   Helpers   ##
+#################
+
+
+def make_filename_valid_for_epub3(filename: str) -> str:
+    filtered_filename = ""
+
+    # Ensure filename contains only allowed chars
+    for char in filename:
+        if char in FILENAME_BANNED_CHARS:
+            continue
+
+        char_allowed = True
+        for range_bottom, range_top in FILENAME_BANNED_CHAR_RANGES:
+            char_ord = ord(char)
+            if char_ord >= range_bottom and char_ord <= range_top:
+                char_allowed = False
+                break
+
+        if char_allowed:
+            filtered_filename += char
+
+    # Ensure filename doesn't end in '.'
+    while len(filtered_filename) > 0 and filtered_filename[-1] == ".":
+        filtered_filename = filtered_filename[:-1]
+
+    if len(filtered_filename) == 0:
+        raise Exception(
+            "Attempted to put file into EPUB with filename containing only invalid characters and/or periods."
+        )
+
+    # Ensure filename is of allowed length, then return
+    if len(filtered_filename.encode("utf-8")) <= 255:
+        return filtered_filename
+    else:
+        # Assumptions:
+        # * File extension exists and is <255 bytes in length
+        # * Filenames have unique numerical identifiers before the 255-byte
+        #   mark, such that truncation won't produce name collisions
+
+        split_filename = filtered_filename.split(".")
+        ext = split_filename[-1]
+        ext_bytes = len(ext.encode("utf-8"))
+        if ext_bytes > 255:
+            raise Exception(
+                "Attempted to put file into EPUB with extension longer than 255 bytes."
+            )
+
+        name_truncated = ".".join(split_filename[:-1])[:-1]
+        full_name_truncated = "%s.%s" % (name_truncated, ext)
+        while len(full_name_truncated.encode("utf-8")) > 255:
```

Comment:
To make sure I'm understanding what's going on here: you're lopping unicode characters off the end one at a time until the length in bytes is less than 255. You're doing this because it's unclear how many characters need to be removed from the end to remove a specific number of bytes. Is that right?

Reply:
Your understanding is correct, yup!